File: frontend/src/api/session-sharing.ts
import { apiClient } from "./client";

export interface SessionShareTokenView {
  id: string;
  token_prefix: string;
  label: string;
  expires_at: string | null;
  max_views: number | null;
  view_count: number;
  last_viewed_at: string | null;
  revoked_at: string | null;
  created_at: string;
}

export interface SessionShareTokenCreated extends SessionShareTokenView {
  token: string;
}

export interface SessionShareMessageView {
  id: string;
  ordinal: number;
  role: string;
  content: string;
  parts: Array<Record<string, unknown>>;
  parent_message_id: string | null;
  created_at: string;
}

export interface SessionShareView {
  id: string;
  title: string;
  scope: string;
  message_count: number;
  status: string;
  created_by: string;
  created_at: string;
}

export interface SessionShareDetailView extends SessionShareView {
  tokens: SessionShareTokenView[];
}

export interface SessionSharePublicView {
  id: string;
  title: string;
  scope: string;
  message_count: number;
  created_at: string;
  messages: SessionShareMessageView[];
}

export interface SessionShareCreate {
  scope: "full" | "range" | "answers";
  from_message_id?: string;
  to_message_id?: string;
  answers_only?: boolean;
  label?: string;
  expires_at?: string | null;
  max_views?: number | null;
}

export function createSessionShare(
  sessionId: string,
  payload: SessionShareCreate,
): Promise<SessionShareTokenCreated> {
  return apiClient.post<SessionShareTokenCreated, SessionShareCreate>(
    `/sessions/${encodeURIComponent(sessionId)}/shares`,
    payload,
  );
}

export function listSessionShares(
  sessionId: string,
): Promise<SessionShareDetailView[]> {
  return apiClient.get<SessionShareDetailView[]>(
    `/sessions/${encodeURIComponent(sessionId)}/shares`,
  );
}

export function revokeSessionShare(shareId: string): Promise<SessionShareView> {
  return apiClient.delete<SessionShareView>(
    `/session-shares/${encodeURIComponent(shareId)}`,
  );
}

export function revokeSessionShareToken(
  shareId: string,
  tokenId: string,
): Promise<SessionShareTokenView> {
  return apiClient.delete<SessionShareTokenView>(
    `/session-shares/${encodeURIComponent(shareId)}/tokens/${encodeURIComponent(tokenId)}`,
  );
}

/** Public (unauthenticated) read-only payload for a shared session. */
export function fetchSharedSession(token: string): Promise<SessionSharePublicView> {
  return apiClient.get<SessionSharePublicView>(
    `/share/${encodeURIComponent(token)}`,
  );
}

/** Frontend route for the public read-only viewer. */
export function sessionShareUrl(rawToken: string): string {
  return `/share/${encodeURIComponent(rawToken)}`;
}

File: backend/app/domain/schemas/session_sharing.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.domain.schemas.common import ORMModel


class SessionShareCreate(BaseModel):
    """Create an immutable session snapshot and a read-only share token.

    ``scope`` selects which messages are frozen into the snapshot:
    - ``full``: every message in the session (default)
    - ``range``: messages between ``from_message_id`` and ``to_message_id``
    - ``answers``: assistant messages only (no user prompts)
    ``answers_only`` is accepted as a synonym for ``scope=answers``.
    """

    scope: str = Field(default="full", pattern="^(full|range|answers)$")
    from_message_id: str | None = None
    to_message_id: str | None = None
    answers_only: bool = False
    label: str = Field(default="", max_length=120)
    expires_at: datetime | None = None
    max_views: int | None = Field(default=None, ge=1)


class SessionShareTokenView(ORMModel):
    id: str
    token_prefix: str
    label: str
    expires_at: datetime | None
    max_views: int | None
    view_count: int
    last_viewed_at: datetime | None
    revoked_at: datetime | None
    created_at: datetime


class SessionShareTokenCreated(SessionShareTokenView):
    """Returned once at creation; the raw token is never persisted."""

    token: str


class SessionShareMessageView(BaseModel):
    id: str
    ordinal: int
    role: str
    content: str
    parts: list[dict]
    parent_message_id: str | None
    created_at: datetime


class SessionShareView(ORMModel):
    id: str
    title: str
    scope: str
    message_count: int
    status: str
    created_by: str
    created_at: datetime


class SessionShareDetailView(SessionShareView):
    tokens: list[SessionShareTokenView] = Field(default_factory=list)


class SessionSharePublicView(BaseModel):
    """Payload served by the unauthenticated public viewer endpoint."""

    id: str
    title: str
    scope: str
    message_count: int
    created_at: datetime
    messages: list[SessionShareMessageView]
